File: Data Structures & Algorithms/redundant-connection/submission-9.cpp
class Solution {
public:
    std::vector<bool> visit;
    std::vector<std::vector<int>> adj;
    std::unordered_set<int> cycle;
    int cycleStart;

    bool DFS(int node, int par)
    {
        if (visit[node])
        {
            cycleStart = node;
            return true;
        }
        visit[node] = true;
        for (int nei : adj[node])
        {
            if (nei == par) continue;
            if (DFS(nei, node))
            {
                if (cycleStart != -1) cycle.insert(node);
                if (node == cycleStart)
                {
                    cycleStart = -1;
                }
                return true;
            }
        }
        return false;
    }

    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int n = edges.size();
        adj.resize(n + 1);
        for (auto& edge : edges)
        {
            int u = edge[0], v = edge[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        visit.resize(n + 1, false);
        cycleStart = -1;
        DFS(1, - 1);

        for (int i = edges.size() - 1; i >= 0; i--)
        {
            int u = edges[i][0], v = edges[i][1];
            if (cycle.count(u) && cycle.count(v))
            {
                return { u, v };
            }
        }

        return {};
    }
};

File: Data Structures & Algorithms/count-paths/submission-9.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        leMap = [[-1] * n for _ in range(m)]
        def DFS(i, j):
            if i >= m or j >= n:
                return 0
            
            if i == m - 1 and j == n - 1:
                return 1 

            if leMap[i][j] != -1:
                return leMap[i][j]

            leMap[i][j] = DFS(i + 1, j) + DFS(i, j + 1)
            return leMap[i][j]
        return DFS(0, 0)
        